File: tools/device-share-hub/android/app/src/main/java/com/zwm/gallery/RemoteRelayTask.java
package com.zwm.gallery;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.IOException;
import java.util.HashSet;
import java.util.Set;

/** Validated view of a relay inbox task. Plain mode carries public file metadata. */
final class RemoteRelayTask {
    static final int MAX_OBJECTS = 1_000;

    final String transferId;
    final String senderDeviceId;
    final String recipientDeviceId;
    final String status;
    final String mode;
    final int objectCount;
    final long totalBytes;
    final java.util.List<ObjectInfo> objects;
    /** Legacy name retained for callers and diagnostics during the protocol migration. */
    final long totalCipherBytes;
    final long expiresAt;

    private RemoteRelayTask(String transferId, String senderDeviceId,
                            String recipientDeviceId, String status,
                            String mode, java.util.List<ObjectInfo> objects,
                            long totalBytes, long expiresAt) {
        this.transferId = transferId;
        this.senderDeviceId = senderDeviceId;
        this.recipientDeviceId = recipientDeviceId;
        this.status = status;
        this.mode = mode;
        this.objects = java.util.Collections.unmodifiableList(objects);
        this.objectCount = objects.size();
        this.totalBytes = totalBytes;
        this.totalCipherBytes = totalBytes;
        this.expiresAt = expiresAt;
    }

    static RemoteRelayTask parse(JSONObject object, String expectedRecipientId,
                                 long nowMs) throws IOException {
        if (object == null) throw new IOException("远程任务为空");
        String transferId = safeId(object.optString("transferId", ""), "远程任务标识");
        String senderDeviceId = safeId(object.optString("senderDeviceId", ""), "发送设备标识");
        String recipientDeviceId = safeId(object.optString("recipientDeviceId", ""), "接收设备标识");
        if (expectedRecipientId == null || !expectedRecipientId.equals(recipientDeviceId)) {
            throw new IOException("远程任务接收设备不匹配");
        }
        String status = object.optString("status", "").trim();
        if (!"ready".equals(status)) throw new IOException("远程任务尚未提交");
        String mode = object.optString("mode", "encrypted").trim();
        if (!"plain".equals(mode) && !"encrypted".equals(mode)) {
            throw new IOException("远程任务传输模式无效");
        }
        long expiresAt = object.optLong("expiresAt", 0L);
        if (expiresAt <= 0L) throw new IOException("远程任务有效期无效");

        JSONArray objects = object.optJSONArray("objects");
        if (objects == null || objects.length() < 1 || objects.length() > MAX_OBJECTS) {
            throw new IOException("远程任务对象数量无效");
        }
        Set<Integer> indexes = new HashSet<>();
        java.util.ArrayList<ObjectInfo> parsedObjects = new java.util.ArrayList<>();
        long totalBytes = 0L;
        for (int i = 0; i < objects.length(); i++) {
            JSONObject item = objects.optJSONObject(i);
            if (item == null) throw new IOException("远程任务对象无效");
            int index = item.optInt("index", -1);
            long bytes = "plain".equals(mode)
                    ? item.optLong("bytes", item.optLong("objectBytes", -1L))
                    : item.optLong("cipherBytes", -1L);
            String sha256 = "plain".equals(mode)
                    ? item.optString("sha256", item.optString("objectSha256", "")).trim()
                    : item.optString("cipherSha256", "").trim();
            String name = item.optString("name", "").trim();
            String mime = item.optString("mime", "application/octet-stream").trim();
            if (index < 0 || !indexes.add(index) || bytes <= 0L
                    || !sha256.matches("[0-9a-fA-F]{64}")
                    || ("plain".equals(mode) && !safeName(name))) {
                throw new IOException("远程对象校验失败");
            }
            if (Long.MAX_VALUE - totalBytes < bytes) {
                throw new IOException("远程任务大小溢出");
            }
            totalBytes += bytes;
            parsedObjects.add(new ObjectInfo(index, bytes, sha256.toLowerCase(),
                    "plain".equals(mode) ? name : "", mime));
        }
        long declaredBytes = "plain".equals(mode)
                ? object.optLong("totalBytes", totalBytes)
                : object.optLong("totalCipherBytes", totalBytes);
        if (declaredBytes != totalBytes) throw new IOException("远程任务总大小不一致");
        return new RemoteRelayTask(transferId, senderDeviceId, recipientDeviceId,
                status, mode, parsedObjects, totalBytes, expiresAt);
    }

    boolean expired(long nowMs) {
        return expiresAt <= nowMs;
    }

    private static String safeId(String value, String label) throws IOException {
        String normalized = value == null ? "" : value.trim();
        if (!normalized.matches("[A-Za-z0-9_-]{8,128}")) {
            throw new IOException(label + "无效");
        }
        return normalized;
    }

    private static boolean safeName(String value) {
        return value != null && !value.isEmpty() && value.length() <= 240
                && !".".equals(value) && !"..".equals(value)
                && value.indexOf('/') < 0 && value.indexOf('\\') < 0
                && value.indexOf('\0') < 0;
    }

    static final class ObjectInfo {
        final int index;
        final long bytes;
        final String sha256;
        final String name;
        final String mime;

        ObjectInfo(int index, long bytes, String sha256, String name, String mime) {
            this.index = index;
            this.bytes = bytes;
            this.sha256 = sha256;
            this.name = name;
            this.mime = mime;
        }
    }
}

File: tools/device-share-hub/ios/Album/RemoteRelayTask.swift
import Foundation

/// Validated view of a relay inbox task. Plain mode carries public file metadata.
struct RemoteRelayTask {
    static let maxObjects = 1_000

    let transferId: String
    let senderDeviceId: String
    let recipientDeviceId: String
    let status: String
    let mode: String
    let contentKind: String
    let objectCount: Int
    let totalBytes: Int64
    let objects: [ObjectInfo]
    /// Legacy name retained while old encrypted tasks are still accepted.
    let totalCipherBytes: Int64
    let expiresAt: Int64

    static func parse(_ object: [String: Any], expectedRecipientId: String,
                      nowMs: Int64 = Int64(Date().timeIntervalSince1970 * 1000)) throws -> RemoteRelayTask {
        guard let transferId = safeId(object["transferId"] as? String),
              let senderDeviceId = safeId(object["senderDeviceId"] as? String),
              let recipientDeviceId = safeId(object["recipientDeviceId"] as? String),
              recipientDeviceId == expectedRecipientId,
              let status = object["status"] as? String, status == "ready",
              let expiresAt = int64(object["expiresAt"]), expiresAt > 0,
              let objects = object["objects"] as? [[String: Any]],
              !objects.isEmpty, objects.count <= maxObjects else {
            throw RemoteRelayError.invalidTask
        }
        let mode = object["mode"] as? String ?? "encrypted"
        guard mode == "plain" || mode == "encrypted" else {
            throw RemoteRelayError.invalidTask
        }
        let contentKind = object["contentKind"] as? String ?? "work"
        guard contentKind == "work" else {
            // iOS can receive works, but it does not silently install mobile
            // update packages. Keep the task visible for a compatible Android
            // device instead of treating an APK as a work file.
            throw RemoteRelayError.invalidTask
        }

        var indexes = Set<Int>()
        var parsedObjects: [ObjectInfo] = []
        var totalBytes: Int64 = 0
        for item in objects {
            guard let index = int(item["index"]), index >= 0,
                  indexes.insert(index).inserted,
                  let bytes = int64(mode == "plain" ? (item["bytes"] ?? item["objectBytes"]) : item["cipherBytes"]),
                  bytes > 0,
                  let hash = (mode == "plain" ? (item["sha256"] ?? item["objectSha256"]) : item["cipherSha256"]) as? String,
                  hash.range(of: "^[0-9a-fA-F]{64}$", options: .regularExpression) != nil,
                  Int64.max - totalBytes >= bytes else {
                throw RemoteRelayError.invalidTask
            }
            let name = item["name"] as? String ?? ""
            guard mode != "plain" || safeName(name) else { throw RemoteRelayError.invalidTask }
            let mime = item["mime"] as? String ?? "application/octet-stream"
            totalBytes += bytes
            parsedObjects.append(ObjectInfo(index: index, bytes: bytes,
                                            sha256: hash.lowercased(), name: name, mime: mime))
        }
        let declaredKey = mode == "plain" ? "totalBytes" : "totalCipherBytes"
        if let declared = int64(object[declaredKey]), declared != totalBytes {
            throw RemoteRelayError.invalidTask
        }
        return RemoteRelayTask(transferId: transferId,
                               senderDeviceId: senderDeviceId,
                               recipientDeviceId: recipientDeviceId,
                               status: status,
                               mode: mode,
                               contentKind: contentKind,
                               objectCount: parsedObjects.count,
                               totalBytes: totalBytes,
                               objects: parsedObjects,
                               totalCipherBytes: totalBytes,
                               expiresAt: expiresAt)
    }

    func expired(at nowMs: Int64) -> Bool { expiresAt <= nowMs }

    private static func safeId(_ value: String?) -> String? {
        guard let value = value?.trimmingCharacters(in: .whitespacesAndNewlines),
              value.count >= 8, value.count <= 128,
              value.range(of: "^[A-Za-z0-9_-]+$", options: .regularExpression) != nil else {
            return nil
        }
        return value
    }

    private static func safeName(_ value: String) -> Bool {
        !value.isEmpty && value.count <= 240 && value != "." && value != ".."
            && !value.contains("/") && !value.contains("\\") && !value.contains("\0")
    }

    struct ObjectInfo {
        let index: Int
        let bytes: Int64
        let sha256: String
        let name: String
        let mime: String
    }

    private static func int(_ value: Any?) -> Int? {
        if let value = value as? Int { return value }
        if let value = value as? NSNumber { return value.intValue }
        return nil
    }

    private static func int64(_ value: Any?) -> Int64? {
        if let value = value as? Int64 { return value }
        if let value = value as? Int { return Int64(value) }
        if let value = value as? NSNumber { return value.int64Value }
        return nil
    }
}
